bokeh/embed.py

```diff
@@ -400,28 +352,24 @@ def autoload_server(plot_object, app_path="/", session_id=DEFAULT_SESSION_ID, ur
 
     return encode_utf8(tag)
 
-def _script_for_render_items(docs_json, render_items, websocket_url,
-                             custom_models, wrap_script=True):
-    # this avoids emitting the "register custom models" code at all
-    # just to register an empty set
-    if (custom_models is not None) and len(custom_models) == 0:
-        custom_models = None
-
-    plot_js = _wrap_in_function(
-        DOC_JS.render(
-            custom_models=custom_models,
-            websocket_url=websocket_url,
-            docs_json=serialize_json(docs_json),
-            render_items=serialize_json(render_items)
+def _script_for_render_items(docs_json, render_items, websocket_url, wrap_script=True):
+    def plot_js(docs_json):
+        return _wrap_in_function(
+            DOC_JS.render(
+                websocket_url=websocket_url,
+                docs_json=serialize_json(docs_json),
+                render_items=serialize_json(render_items),
+            )
         )
-    )
+
     if wrap_script:
-        return SCRIPT_TAG.render(js_code=plot_js)
+        id = str(uuid.uuid4())
+        return SCRIPT_TAG.render(docs_json_id=id, docs_json=serialize_json(docs_json), js_code=plot_js(id))
     else:
-        return plot_js
+        return plot_js(docs_json)
```

Comment:
I think this refactor is incomplete and we end up with one function that really does two separate things. As it is, the function is named and expected elsewhere to return one hunk of script, which is optionally wrapped in one `<script>` tag. If we want to change it so that to embed render items, you need N script tags, we should rename the function `_scripts_for_render_items`, return a list, and have all callers handle the list result.

`wrap_script` should control whether each element is wrapped in a `<script>` tag, not how many elements there are. Though with docs_json in its own tag I think we have to always wrap_script so that parameter would maybe have to go.

The simplest thing for this PR is probably to drop the "externalize docs json into its own <script> tag" and do that as its own PR, but if you think externalizing the docs json is important to this PR, I think we should finish the refactor - do it always/consistently. There's no value I see to having two ways to do this: either externalize the docs_json into tags, or don't.

Going from one script to a list of scripts will break the `components()` API I think. If we don't want to break `components()` I'm not sure we should do this at all, is it really worth two codepaths?

If we _do_ have two codepaths, I'd rather not overload `wrap_script` - it refers to whether to wrap the scripts in a script tag, which is separate from whether the docs_json is in its own script tag. Maybe a separate function `_scripts_for_render_items_with_externalized_json` ?